app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/endpoints.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/keycode_state_changed.h>
+#include <zmk/events/modifiers_state_changed.h>
+#include <zmk/hid.h>
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+enum mode {
+    MODE_KEY_UP = 0,
+    MODE_KEY_DOWN = 1,
+    MODE_HOLD = 2,
+};
+
+struct behavior_macro_config {
+    int mode;
+    int sleep;
+    int behavior_count;
+    struct zmk_behavior_binding *behaviors;
+};
+
+struct behavior_macro_data {};
+
+static int behavior_macro_init(const struct device *dev) { return 0; }
+
+static void press_and_release_combo(const struct behavior_macro_config *cfg,
+                                    struct zmk_behavior_binding_event event) {
+    for (int index = 0; index < cfg->behavior_count; index++) {
+        const struct device *behavior = device_get_binding(cfg->behaviors[index].behavior_dev);
+        if (!behavior) {
+            break;
+        }
+        behavior_keymap_binding_pressed(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
```

Comment:
This theoretically ties up the system work queue/thread, especially for longer macros.. how well did this work in real life for longer ones? Dropped messages for the HOG queue?

Reply:
I haven't tested it on long macros nor in real life. 

The assumption now is that macros will not permit any other keyboard interaction until they have finished executing. If that means some queues get full, I suggest people increase their queues as a workaround?